File: src/l10n.ts
import { readFileSync, existsSync } from 'fs';
import { join } from 'path';
import { Locale } from 'discord.js';
import { ZENOX_LOCALES, DatabaseKey, LOCALE_TO_TEXTMAP } from './static/constants';
import { Game } from './static/enums';

const SOURCE_LANG = "en-US";
const L10N_PATH = "./zenox/l10n";
const SUPPORTED_LANG = Object.values(ZENOX_LOCALES).map(x => x.value);

export function genStringKey(string: string): string {
  return string.replace(/ /g, "_").replace(/,/g, "").replace(/\./g, "").replace(/-/g, "_").toLowerCase();
}

export interface LocaleStrOptions {
  key?: string;
  customStr?: string;
  translate?: boolean;
  dataGame?: Game;
  [key: string]: any;
}

export class LocaleStr {
  key?: string;
  customStr?: string;
  extras: Record<string, any>;
  translate_: boolean;
  game?: Game;

  constructor(options: LocaleStrOptions) {
    this.key = options.key;
    this.customStr = options.customStr;
    this.extras = { ...options };
    delete this.extras.key;
    delete this.extras.customStr;
    delete this.extras.translate;
    delete this.extras.dataGame;
    this.translate_ = options.translate ?? true;
    this.game = options.dataGame;
  }

  translate(locale: Locale): string {
    return translator.translate(this, locale);
  }
}

export class Translator {
  private _localizations: Record<string, Record<string, string>> = {};
  private _gameTextmaps: Record<Game, Record<Locale, Record<string, string>>> = {
    [Game.GENSHIN]: {},
    [Game.STARRAIL]: {},
    [Game.ZZZ]: {}
  };

  constructor() {
    this.loadL10nFiles();
    this.loadTextMaps();
  }

  private loadTextMaps(): void {
    for (const game of Object.values(Game)) {
      const gameStr = DatabaseKey[game];
      for (const locale of Object.keys(ZENOX_LOCALES)) {
        const fileStr = LOCALE_TO_TEXTMAP[locale as Locale];
        this._gameTextmaps[game][locale as Locale] = this.readJson(
          `zenox/bot/data/text_maps/${gameStr}/${fileStr}.json`
        );
      }
    }
  }

  private loadL10nFiles(): void {
    // This would need to be implemented to read YAML files
    // For now, we'll create a basic structure
    this._localizations[SOURCE_LANG] = {};
  }

  private readJson(filepath: string): Record<string, string> {
    try {
      if (!existsSync(filepath)) {
        return {};
      }
      const jsonData = JSON.parse(readFileSync(filepath, 'utf8'));
      const localizations: Record<string, string> = {};
      for (const [locale, strings] of Object.entries(jsonData)) {
        localizations[locale] = strings as string;
      }
      return localizations;
    } catch (error) {
      console.error(`Error reading JSON file ${filepath}:`, error);
      return {};
    }
  }

  private readYaml(filepath: string): Record<string, string> {
    // This would need to be implemented with a YAML parser
    // For now, return empty object
    return {};
  }

  private translateExtras(extras: Record<string, any>, locale: Locale): Record<string, any> {
    const extras_: Record<string, any> = {};
    for (const [k, v] of Object.entries(extras)) {
      if (v instanceof LocaleStr) {
        extras_[k] = this.translate(v, locale);
      } else if (Array.isArray(v) && v[0] instanceof LocaleStr) {
        extras_[k] = v.map(i => this.translate(i, locale)).join("/");
      } else {
        extras_[k] = v;
      }
    }
    return extras_;
  }

  private static getStringKey(string: LocaleStr): string {
    if (string.key === undefined) {
      if (string.customStr === undefined) {
        throw new Error("Either key or customStr must be provided.");
      }
      return genStringKey(string.customStr);
    }
    return string.key;
  }

  translate(string: LocaleStr | string, locale: Locale): string {
    if (typeof string === "string") {
      return string;
    }

    const extras = this.translateExtras(string.extras, locale);
    const stringKey = Translator.getStringKey(string);

    let sourceString: string | undefined;

    if (string.game !== undefined) {
      sourceString = this._gameTextmaps[string.game][Locale.EnglishUS]?.[stringKey];
    } else {
      sourceString = this._localizations[SOURCE_LANG]?.[stringKey];
    }

    if (string.translate_ && sourceString === undefined && string.customStr === undefined) {
      throw new Error(`String '${stringKey}' not found in source language file`);
    }

    let translation: string | undefined;

    if (string.game !== undefined) {
      translation = this._gameTextmaps[string.game][locale]?.[stringKey];
    } else {
      translation = this._localizations[locale.value]?.[stringKey];
    }

    translation = translation || sourceString || string.customStr || stringKey;

    try {
      translation = translation.replace(/\{(\w+)\}/g, (match, key) => {
        return extras[key] !== undefined ? extras[key] : match;
      });
    } catch (error) {
      // Ignore formatting errors
    }

    return translation;
  }
}

export class AppCommandTranslator {
  constructor() {
    // Discord.js doesn't have built-in translator like discord.py
    // This would need to be implemented differently
  }

  async translate(
    string: any,
    locale: Locale,
    context: any
  ): Promise<string> {
    const key = string.extras?.key;
    if (key === undefined) {
      return string.message;
    }
    return translator.translate(new LocaleStr({ key }), locale);
  }
}

export const translator = new Translator();

File: zenox/l10n.py
import pathlib
import contextlib
from pathlib import Path
import os
import json, yaml
import discord
from discord import app_commands
from typing import Self, Any
from .static.constants import ZENOX_LOCALES, DatabaseKey, LOCALE_TO_TEXTMAP
from .static.enums import Game

from discord.enums import Locale

SOURCE_LANG = "en-US"
L10N_PATH = pathlib.Path("./zenox/l10n")
SUPPORTED_LANG = [x.value for x in ZENOX_LOCALES]

def gen_string_key(string: str) -> str:
    return string.replace(" ", "_").replace(",", "").replace(".", "").replace("-", "_").lower()

class LocaleStr:
    def __init__(
        self,
        *,
        key: str | None = None,
        custom_str: str | None = None,
        translate: bool = True,
        data_game: Game | None = None,
        **kwargs
    ) -> None:
        self.key = key
        self.custom_str = custom_str
        self.extras: dict[str, Any] = kwargs
        self.translate_ = translate
        self.game = data_game
    
    def translate(self, locale: Locale) -> str:
        return translator.translate(self, locale)

class Translator:
    def __init__(self) -> None:
        self._localizations: dict[str, dict[str, str]] = {}
        self._game_textmaps: dict[Game, dict[str, dict[str, str]]] = {
            game: {} for game in Game
        }
        self.load_l10n_files()
        self.load_text_maps()

    def load_text_maps(self) -> None:
        for game in Game:
            game_str = DatabaseKey[game]
            for locale in ZENOX_LOCALES:
                file_str = LOCALE_TO_TEXTMAP[locale]
                self._game_textmaps[game][locale] = self.read_json(
                    f"zenox/bot/data/text_maps/{game_str}/{file_str}.json"
                )

    def load_l10n_files(self) -> None:
        for filepath in L10N_PATH.glob("*.yaml"):
            if not filepath.exists():
                continue
            lang = filepath.stem
            self._localizations[lang] = self.read_yaml(filepath.as_posix())

    def read_json(self, filepath: str) -> dict[str, str]:
        with open(filepath, 'r', encoding='utf-8') as f:
            json_data = json.load(f)
        localizations = {}
        for locale, strings in json_data.items():
            localizations[locale] = strings
        return localizations

    def read_yaml(self, filepath: Path) -> None:
        with open(filepath, 'r', encoding='utf-8') as f:
            yaml_data = yaml.safe_load(f)
        localizations = {}
        for locale, strings in yaml_data.items():
            localizations[locale] = strings
        return localizations
    
    def _translate_extras(self, extras: dict[str, Any], locale: Locale) -> dict[str, Any]:
        extras_: dict[str, Any] = {}
        for k, v in extras.items():
            if isinstance(v, LocaleStr):
                extras_[k] = self.translate(v, locale)
            elif isinstance(v, list) and isinstance(v[0], LocaleStr):
                extras_[k] = "/".join([self.translate(i, locale) for i in v])
            else:
                extras_[k] = v
        return extras_
    
    @staticmethod
    def _get_string_key(string: LocaleStr) -> str:
        if string.key is None:
            if string.custom_str is None:
                raise ValueError("Either key or custom_str must be provided.")
            return gen_string_key(string.custom_str)
        return string.key

    def translate(self, string: LocaleStr | str, locale: Locale):
        if (isinstance(string, str)):
            return string
        
        extras = self._translate_extras(string.extras, locale)
        string_key = self._get_string_key(string)

        if string.game is not None:
            source_string = self._game_textmaps[string.game][discord.Locale(SOURCE_LANG)].get(string_key)

        else:
            source_string = self._localizations[SOURCE_LANG].get(string_key)
        
        if string.translate_ and source_string is None and string.custom_str is None:
            raise ValueError(f"String '{string_key!r}' not found in source language file")
        
        if string.game is not None:
            translation = self._game_textmaps[string.game][locale].get(string_key)
        else:
            translation = self._localizations.get(locale.value, {}).get(string_key)
        
        translation = translation or source_string or string.custom_str or string_key

        with contextlib.suppress(KeyError):
            translation = translation.format(**extras)
        
        return translation

class AppCommandTranslator(app_commands.Translator):
    def __init__(self) -> None:
        super().__init__()

    async def translate(
        self,
        string: app_commands.locale_str,
        locale: discord.Locale,
        context: app_commands.TranslationContext,
    ) -> str:
        if (key := string.extras.get("key")) is None:
            return string.message
        return translator.translate(LocaleStr(key=key), locale)
    
translator = Translator()